checker/src/callbacks.rs

```diff
@@ -372,4 +387,51 @@ impl MiraiCallbacks {
         });
         mir_visitor.visit_body(&name, &[]);
     }
+
+    /// Extract test functions from the promoted constants of a test runner main function.
+    ///
+    /// Currently, the #[test] attribute generates code like this:
+    ///  
+    ///     extern crate test;
+    ///
+    ///     pub fn main() -> () {
+    ///       test::test_main_static(&[&test, ...)...])
+    ///     }
+    ///
+    ///     pub const test: test:TestDescAndFn = TestDecAndFn{
+    ///       ...,
+    ///       testfn: test::StaticTestFn(|| test::assert_test_result(test())
+    ///     };
+    ///
+    ///     pub fn test() { <original user test code> }
+    ///  
+    /// We can thus find the names (but not def def_id's) of the test functions in the main
+    /// method (via const test in the example). However, the constant slice in main will be promoted
+    /// into a constant initializer function in MIR, so we need to look there.
+    /// We therefore iterate overall promoted functions belonging to the main function, looking for
+    /// a statement like `_n = const <test name>` which loads the constant for a given test.
+    ///
+    /// This method is indeed not very stable and may break on changes to the compilation
+    /// scheme or the test framework.
+    fn extract_test_fns(tcx: TyCtxt<'_>, def_id: DefId) -> Vec<String> {
+        let mut result = vec![];
+        for body in tcx.promoted_mir(def_id).iter() {
```

Comment:
This should be optimized_mir

Reply:
promoted_mir is something else AFAICT. It delivers all the promoted function's mir as an iterator, not the mir of this function. I have not seen anything like promoted_optimized_mir.

Reply:
This is probably fine. It returns blocks of MIR that created promoted constants. A promoted constant is something that needs an address. Apparently this is where the test functions live.